Assets/Scripts/PlayerSupervisor.cs

```diff
@@ -1,317 +1,385 @@
+    private int boundaryHits = 0;
+
+    private int paddleHits = 0;
+
+    private GameObject trainingBlocksInstance;
+    private int activeBlocks;
+    private PlayerState playerState = PlayerState.Waiting;
+
+    [Header("Game Environment Settings")]
+
+    public float minPaddlePosX = 1f;
+    public float maxPaddlePosX = 15f;
+    public float instanceWidth = 16f;
+    public float instanceHeight = 12f;
+   
+    [Range(10f, 200f)]
+    public float paddleMoveSpeed = 100f;
+
+    [Range(.1f, 10f)]
+    public float moveStep = 2f;
+
+    public int boundaryReboundLimit = 10;
+
+    [Tooltip("How often to check for anomalies (0 eliminates check)")]
+    [Range(0f, 2f)]
+    public float detectionFreq = 1f;
+
+    [Tooltip("Angle that ball will ricochet off ceiling to prevent juggling")]
+    [Range(0f, 5f)]
+    public float ceilingReboundAngle = 0f;
+
+    // Start is called before the first frame update
+    void Start()
+    {
+        gameManager = FindObjectOfType<GameManager>();
+        
+        if (!playerAgent)
+            playerAgent = FindObjectOfType<PlayerAgent>();
+
+        // the code will check whether or not to execute
+        // based on the block.name assigned in the Inspector Window
+        // in the RandomBlockCreator empty child object
+        randomBlockCreator = FindObjectOfType<RandomBlockCreator>();
+        if (randomBlockCreator)
+            randomBlockCreator.setupBlocks();
+        
+        if (!ball)
+            ball = FindObjectOfType<Ball>();
+        ballOffset = ball.transform.localPosition;
+
+        // Performance tracking - has to come before countblocks
+        if(!isMultiTraining && gameManager.trackingPerformanceTF)
+            dataManager = FindObjectOfType<PerformanceDataManager>();
+        
+        if (!paddle)
+            paddle = FindObjectOfType<Paddle>();
+        paddleOffset = paddle.transform.localPosition;
+
+        // Check if scene is ready for training
+        if (gameManager.trainingMode)
+            ResetState();
+        else
+            CountBlocks();
+    }
+
+    void CountBlocks()
+    {
+        activeBlocks = 0;
+
+        // TODO need to count only blocks in each player's environment to
+        //  work for general multiplayer use; Maybe keep all blocks in a
+        //  container for easier tracking
+        if (gameManager.trainingMode)
+        {
+            foreach(Transform child in trainingBlocksInstance.transform)
+            {
+                if (child.gameObject.GetComponent<Block>() && child.gameObject.activeSelf)
+                    ++activeBlocks;
+            }
+        }
+        else
+        {
+            foreach (Block block in FindObjectsOfType<Block>())
+            {
+                if (block.gameObject.activeSelf)
+                    ++activeBlocks;
+            }
+        }
+        if(!isMultiTraining && gameManager.trackingPerformanceTF)
+            dataManager.SetStartingNumBlocks(activeBlocks);
```

Comment:
Setting dataManager.startingNumBlocks makes me a little nervous here - currently our solution only uses CountBlocks() when getting an initial block count at the beginning of the scene, but there's nothing to really stop it from being called at other times when it would be incorrect to run SetStartingNumBlocks(). I think this is fine as long as well agree that CountBlocks is only for taking an initial count when a new game starts. But you actually don't need this call here at all since a few methods below in StartGame the SetStartingNumBlocks call already sets the value for DataManager (as long as the agent eventually launches the ball).